msbuild/Xamarin.MacDev.Tasks.Core/Tasks/IntentDefinitionCodegenTaskBase.cs

```diff
@@ -0,0 +1,439 @@
+﻿using System;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Collections.Generic;
+
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+namespace Xamarin.MacDev.Tasks
+{
+	public abstract class IntentDefinitionCodegenTaskBase : Task
+	{
+		static readonly string[] DefaultResponseCodes = {
+			"Unspecified", "Ready", "ContinueInApp", "InProgress", "Success", "Failure", "FailureRequiringAppLaunch"
+		};
+
+		#region Inputs
+
+		public string SessionId { get; set; }
+
+		[Required]
+		public ITaskItem IntentDefinition { get; set; }
+
+		[Required]
+		public string IntermediateOutputPath { get; set; }
+
+		[Required]
+		public string RootNamespace { get; set; }
+
+		#endregion
+
+		#region Outputs
+
+		[Output]
+		public ITaskItem[] GeneratedSources { get; set; }
+
+		#endregion
+
+		static string GetCSharpName (string name)
+		{
+			if (name.Length > 1)
+				return char.ToUpperInvariant (name[0]) + name.Substring (1);
+
+			return name.ToUpperInvariant ();
+		}
+
+		static string GetCSharpTypeName (string type)
+		{
+			switch (type) {
```

Comment:
similarly, you can you `nameof()` on type names here: `return nameof (NSString);`